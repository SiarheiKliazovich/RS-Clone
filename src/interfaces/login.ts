import { IUser } from "./user";
// export type User = {
//   email: string;
//   role: string;
// };

export interface ILogin {
  user: IUser;
  isLogin: boolean;
}
