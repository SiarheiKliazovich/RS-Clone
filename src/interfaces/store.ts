export interface IStore {
  appInterface: IAppInterface;
}

export interface IAppInterface {
  isSidebarFixed: boolean;
  isNightMode: boolean;
  isSidebarAccentMode: boolean;
  isNavbarNightMode: boolean;
  isProfileShow: boolean;
  isSidebarShow: boolean;
  accentColor: {
    static: string;
    hover: string;
  };
}

export interface IAppInterfaceLocalStorage {
  isSidebarFixed: boolean;
  isNightMode: boolean;
  isSidebarAccentMode: boolean;
  isNavbarNightMode: boolean;
  isProfileShow?: boolean;
  isSidebarShow?: boolean;
  accentColor: {
    static: string;
    hover: string;
  };
}

export enum ActionTypes {
  updateFixedSidebar = "UPDATE_FIXED_SIDEBAR",
  updateNightMode = "UPDATE_NIGHT_MODE",
  updateSidebarAccentMode = "UPDATE_SIDEBAR_ACCENT_MODE",
  updateAccentColor = "UPDATE_ACCENT_COLOR",
  updateNavbarNightMode = "UPDATE_NAVBAR_NIGHT_MODE",
  showProfile = "SHOW_PROFILE",
  showSidebar = "SHOW_SIDEBAR",
  resetInterfaceToDefault = "RESET_INTERFACE_TO_DEFAULT",
}
