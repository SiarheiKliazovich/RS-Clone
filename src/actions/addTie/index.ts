import { ActionTypes } from "../../interfaces/store";
import { ITie } from "../../interfaces/tie";

export const addTieFetching = () => {
  return {
    type: ActionTypes.addTieFetching,
  };
};

export const addTieFetched = (tie: ITie) => {
  return {
    type: ActionTypes.addTieFetched,
    payload: tie,
  };
};

export const addTieFetchingError = () => {
  return {
    type: ActionTypes.addTieFetchingError,
  };
};
