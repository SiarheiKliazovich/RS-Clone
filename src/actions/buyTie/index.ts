import { IOrder } from "../../interfaces/order";
import { ActionTypes } from "../../interfaces/store";

export const buyTieFetching = () => {
  return {
    type: ActionTypes.buyTieFetching,
  };
};

export const buyTieFetched = (order: IOrder) => {
  return {
    type: ActionTypes.buyTieFetched,
    payload: order,
  };
};

export const buyTieFetchingError = () => {
  return {
    type: ActionTypes.buyTieFetchingError,
  };
};
