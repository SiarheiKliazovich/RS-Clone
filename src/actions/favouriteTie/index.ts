import { IOrder } from "../../interfaces/order";
import { ActionTypes } from "../../interfaces/store";

export const favouriteTieFetching = () => {
  return {
    type: ActionTypes.favouriteTieFetching,
  };
};

export const favouriteTieFetched = (order: IOrder) => {
  return {
    type: ActionTypes.favouriteTieFetched,
    payload: order,
  };
};

export const favouriteTieFetchingError = () => {
  return {
    type: ActionTypes.favouriteTieFetchingError,
  };
};
