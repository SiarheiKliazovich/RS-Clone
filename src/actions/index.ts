import { ActionTypes } from "../interfaces/store";
import { ILogin } from "../interfaces/login";
import { IUser } from "../interfaces/user";
import { IOrder } from "../interfaces/order";
import { ILang } from "../interfaces/lang";
import { ITie } from "../interfaces/tie";

export const updateFixedSidebar = (isSidebarFixed: boolean) => ({
  type: ActionTypes.updateFixedSidebar,
  payload: isSidebarFixed,
});

export const updateNightMode = (isNightMode: boolean) => ({
  type: ActionTypes.updateNightMode,
  payload: isNightMode,
});

export const updateSidebarAccentMode = (isSidebarAccentMode: boolean) => ({
  type: ActionTypes.updateSidebarAccentMode,
  payload: isSidebarAccentMode,
});

export const updateNavbarNightMode = (isNavbarNightMode: string) => ({
  type: ActionTypes.updateNavbarNightMode,
  payload: isNavbarNightMode,
});

export const updateAccentColor = (accentColor: {
  static: string;
  hover: string;
}) => ({
  type: ActionTypes.updateAccentColor,
  payload: accentColor,
});

export const resetInterfaceToDefault = () => ({
  type: ActionTypes.resetInterfaceToDefault,
});

export const showProfile = () => ({ type: ActionTypes.showProfile });

export const showSidebar = () => ({ type: ActionTypes.showSidebar });

export const registration = () => ({ type: ActionTypes.registration });

export const authorization = (user: ILogin, isLogin: boolean) => ({
  type: ActionTypes.authorization,
  payload: { user, isLogin },
});

export const usersFetching = () => {
  return {
    type: ActionTypes.usersFetching,
  };
};

export const usersFetched = (users: IUser[]) => {
  return {
    type: ActionTypes.usersFetched,
    payload: users,
  };
};

export const usersFetchingError = () => {
  return {
    type: ActionTypes.usersFetchingError,
  };
};

export const userDeleted = (id: string) => {
  return {
    type: ActionTypes.userDeleted,
    payload: id,
  };
};

export const ordersFetching = () => {
  return {
    type: ActionTypes.ordersFetching,
  };
};

export const ordersFetched = (orders: IOrder[]) => {
  return {
    type: ActionTypes.ordersFetched,
    payload: orders,
  };
};

export const ordersFetchingError = () => {
  return {
    type: ActionTypes.ordersFetchingError,
  };
};

export const orderDeleted = (id: string) => {
  return {
    type: ActionTypes.orderDeleted,
    payload: id,
  };
};

export const changeLanguage = (lang: ILang) => {
  return {
    type: ActionTypes.changeLanguage,
    payload: lang,
  };
};

export const myTiesFetching = () => {
  return {
    type: ActionTypes.myTiesFetching,
  };
};

export const myTiesFetched = (ties: ITie[]) => {
  return {
    type: ActionTypes.myTiesFetched,
    payload: ties,
  };
};

export const myTiesFetchingError = () => {
  return {
    type: ActionTypes.myTiesFetchingError,
  };
};

export const myTieDeleted = (id: string) => {
  return {
    type: ActionTypes.myTieDeleted,
    payload: id,
  };
};

export const myOrdersFetching = () => {
  return {
    type: ActionTypes.myOrdersFetching,
  };
};

export const myOrdersFetched = (orders: IOrder[]) => {
  return {
    type: ActionTypes.myOrdersFetched,
    payload: orders,
  };
};

export const myOrdersFetchingError = () => {
  return {
    type: ActionTypes.myOrdersFetchingError,
  };
};

export const myOrderDeleted = (id: string) => {
  return {
    type: ActionTypes.myOrderDeleted,
    payload: id,
  };
};
