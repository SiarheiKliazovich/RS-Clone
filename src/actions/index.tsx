import { ActionTypes } from "../interfaces/store";
import { ILogin } from "../interfaces/login";
import { IUser } from "../interfaces/user";
import { IOrder } from "../interfaces/order";
import { ILang } from "../interfaces/lang";

export const updateFixedSidebar = (isSidebarFixed: boolean) => ({
  type: ActionTypes.updateFixedSidebar,
  payload: isSidebarFixed,
});

export const updateNightMode = (isNightMode: boolean) => ({
  type: ActionTypes.updateNightMode,
  payload: isNightMode,
});

export const updateSidebarAccentMode = (isSidebarAccentMode: boolean) => ({
  type: ActionTypes.updateSidebarAccentMode,
  payload: isSidebarAccentMode,
});

export const updateNavbarNightMode = (isNavbarNightMode: string) => ({
  type: ActionTypes.updateNavbarNightMode,
  payload: isNavbarNightMode,
});

export const updateAccentColor = (accentColor: {
  static: string;
  hover: string;
}) => ({
  type: ActionTypes.updateAccentColor,
  payload: accentColor,
});

export const resetInterfaceToDefault = () => ({
  type: ActionTypes.resetInterfaceToDefault,
});

export const showProfile = () => ({ type: ActionTypes.showProfile });

export const showSidebar = () => ({ type: ActionTypes.showSidebar });

export const registration = () => ({ type: ActionTypes.registration });

export const authorization = (user: ILogin, isLogin: boolean) => ({
  type: ActionTypes.authorization,
  payload: { user, isLogin },
});

export const usersFetching = () => {
  return {
    type: ActionTypes.usersFetching,
  };
};

export const usersFetched = (users: IUser[]) => {
  return {
    type: ActionTypes.usersFetched,
    payload: users,
  };
};

export const usersFetchingError = () => {
  return {
    type: ActionTypes.usersFetchingError,
  };
};

export const userDeleted = (id: string) => {
  return {
    type: ActionTypes.userDeleted,
    payload: id,
  };
};

export const ordersFetching = () => {
  return {
    type: ActionTypes.ordersFetching,
  };
};

export const ordersFetched = (orders: IOrder[]) => {
  return {
    type: ActionTypes.ordersFetched,
    payload: orders,
  };
};

export const ordersFetchingError = () => {
  return {
    type: ActionTypes.ordersFetchingError,
  };
};

export const orderDeleted = (id: string) => {
  return {
    type: ActionTypes.orderDeleted,
    payload: id,
  };
};

export const changeLanguage = (lang: ILang) => {
  return {
    type: ActionTypes.changeLanguage,
    payload: lang,
  };
};
