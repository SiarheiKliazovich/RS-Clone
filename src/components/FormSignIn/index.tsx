import { FunctionComponent } from "react";
import { Link, useNavigate } from "react-router-dom";
import { Formik, Field, Form, ErrorMessage } from "formik";
import * as Yup from "yup";
import { useDispatch, useSelector } from "react-redux";
import { useState, useEffect } from "react";
import { useTranslation } from "react-i18next";

import { IUser } from "../../interfaces/user";
import { login } from "../../services/apiAuth";
import { authorization } from "../../actions";
import { IStore } from "../../interfaces/store";
import { nightTheme } from "../../data/constants";

const FormSignIn: FunctionComponent = () => {
  const [error, setError] = useState<string>("");

  const dispatch = useDispatch();
  const navigate = useNavigate();

  useEffect(() => {
    setTimeout(() => setError(""), 5000);
  }, [error]);

  const { t } = useTranslation("dataLang");

  const click = async (body: Pick<IUser, "email" | "password">) => {
    try {
      const data = await login(body);
      dispatch(authorization(data, true));
      if (data) {
        if (data.role === "ADMIN" || data.role === "MANAGER") {
          setTimeout(() => navigate("/dashboard"), 2000);
        } else {
          setTimeout(() => navigate("/"), 2000);
        }
      }
    } catch (e) {
      setError(e as string);
    }
  };

  const interfaceSettings = useSelector((state: IStore) => state.appInterface);

  const { accentColor, isNavbarNightMode } = interfaceSettings;
  const backgroundColor = nightTheme.background.element;

  return (
    <div className="sign__content">
      <h2 className="sign__title">{t("sign.signIn")}</h2>
      <div className="sign__text">{t("sign.text-1")}</div>
      <Formik
        initialValues={{
          email: "",
          password: "",
        }}
        validationSchema={Yup.object({
          email: Yup.string()
            .email("Incorrect email")
            .required("Required field"),
          password: Yup.string()
            .min(4, "Minimum 4 symbols")
            .max(10, "Maximum 10 symbols")
            .required("Required field"),
        })}
        onSubmit={(values) => {
          click(values);
        }}
      >
        <Form className="form">
          <Field
            className="field"
            id="email"
            name="email"
            type="email"
            placeholder="Email"
          />
          <ErrorMessage className="error" name="email" component="div" />
          <Field
            className="field"
            id="password"
            name="password"
            type="password"
            placeholder="Password"
          />
          <ErrorMessage className="error" name="password" component="div" />
          <button
            style={{
              backgroundColor: isNavbarNightMode
                ? backgroundColor
                : accentColor.static,
            }}
            className="sign__button"
            type="submit"
          >
            {t("sign.signIn")}
          </button>
          {error && (
            <div className="error-tooltip">
              <i className="fa fa-warning" /> {error}
            </div>
          )}
        </Form>
      </Formik>
      <div className="sign__text-link">
        {t("sign.create")}{" "}
        <Link
          style={{
            color: isNavbarNightMode ? backgroundColor : accentColor.static,
          }}
          className="text-link"
          to="/sign-up"
        >
          {t("sign.signUp")}
        </Link>
      </div>
    </div>
  );
};

export default FormSignIn;
