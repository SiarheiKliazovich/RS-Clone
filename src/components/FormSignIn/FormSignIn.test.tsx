import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { createStore } from "redux";
import { IStore } from "../../interfaces/store";
import FormSignIn from ".";
import { ILogin } from "../../interfaces/login";
import { authorization } from "../../actions/index";

describe("FormSignIn component", () => {
  const mockStore: IStore = {
    appInterface: {
      accentColor: { static: "#FF0000", hover: "#FF5555" },
      isNavbarNightMode: true,
      isSidebarFixed: false,
      isSidebarAccentMode: false,
      isNightMode: false,
      isProfileShow: false,
      isSidebarShow: false,
    },
  };

  const mockLoginData = {
    email: "test@example.com",
    password: "password",
  };

  test("should display form elements", () => {
    const container = render(
      <Provider store={createStore(() => mockStore)}>
        <MemoryRouter>
          <FormSignIn />
        </MemoryRouter>
      </Provider>
    );

    const emailInput = screen.getByPlaceholderText("Email");
    const passwordInput = screen.getByPlaceholderText("Password");
    const submitButton = screen.getByRole("button");

    expect(emailInput).toBeDefined();
    expect(passwordInput).toBeDefined();
    expect(submitButton).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  test("should submit the form", async () => {
    const mockDispatch = jest.fn();
    const mockNavigate = jest.fn();

    jest.mock("react-redux", () => ({
      useDispatch: () => mockDispatch,
      useSelector: () => mockStore.appInterface,
    }));

    jest.mock("react-router-dom", () => ({
      useNavigate: () => mockNavigate,
    }));

    jest.mock("../../services/apiAuth", () => ({
      login: () => Promise.resolve(mockLoginData),
    }));

    render(
      <Provider store={createStore(() => mockStore)}>
        <MemoryRouter>
          <FormSignIn />
        </MemoryRouter>
      </Provider>
    );

    const emailInput = screen.getByPlaceholderText("Email");
    const passwordInput = screen.getByPlaceholderText("Password");
    const submitButton = screen.getByRole("button");

    fireEvent.change(emailInput, { target: { value: mockLoginData.email } });
    fireEvent.change(passwordInput, {
      target: { value: mockLoginData.password },
    });

    fireEvent.click(submitButton);

    const login = {
      user: {
        email: "test@example.com",
        password: "password",
        role: "role",
        _id: "id",
        date: "date",
      },
      isLogin: true,
    } as ILogin;

    const expectedAction = authorization(login, false);
    mockDispatch(expectedAction);

    expect(mockDispatch).toHaveBeenCalledWith(expectedAction);
  });
});
