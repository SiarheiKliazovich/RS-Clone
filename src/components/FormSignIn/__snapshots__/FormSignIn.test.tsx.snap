// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSignIn component should display form elements 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sign__content"
      >
        <h2
          class="sign__title"
        >
          sign.signIn
        </h2>
        <div
          class="sign__text"
        >
          sign.text-1
        </div>
        <form
          action="#"
          class="form"
        >
          <input
            class="field"
            id="email"
            name="email"
            placeholder="Email"
            type="email"
            value=""
          />
          <input
            class="field"
            id="password"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <button
            class="sign__button"
            style="background-color: rgb(43, 52, 59);"
            type="submit"
          >
            sign.signIn
          </button>
        </form>
        <div
          class="sign__text-link"
        >
          sign.create
           
          <a
            class="text-link"
            href="/sign-up"
            style="color: rgb(43, 52, 59);"
          >
            sign.signUp
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sign__content"
    >
      <h2
        class="sign__title"
      >
        sign.signIn
      </h2>
      <div
        class="sign__text"
      >
        sign.text-1
      </div>
      <form
        action="#"
        class="form"
      >
        <input
          class="field"
          id="email"
          name="email"
          placeholder="Email"
          type="email"
          value=""
        />
        <input
          class="field"
          id="password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
        <button
          class="sign__button"
          style="background-color: rgb(43, 52, 59);"
          type="submit"
        >
          sign.signIn
        </button>
      </form>
      <div
        class="sign__text-link"
      >
        sign.create
         
        <a
          class="text-link"
          href="/sign-up"
          style="color: rgb(43, 52, 59);"
        >
          sign.signUp
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
